src/figuretype/depot.c

```diff
@@ -0,0 +1,395 @@
+    }
+
+    switch (f->action_state) {
+        case FIGURE_ACTION_239_DEPOT_CART_PUSHER_HEADING_TO_SOURCE:
+        case FIGURE_ACTION_240_DEPOT_CART_PUSHER_AT_SOURCE:
+        {
+            if (f->destination_building_id == order->src_storage_id) {
+                return 1;
+            } else {
+                return 0;
+            }
+            break;
+        }
+        case FIGURE_ACTION_241_DEPOT_CART_HEADING_TO_DESTINATION:
+        case FIGURE_ACTION_242_DEPOT_CART_PUSHER_AT_DESTINATION:
+        {
+            if (f->resource_id == order->resource_type
+                && f->destination_building_id == order->dst_storage_id
+                && f->loads_sold_or_carrying > 0) {
+                return 1;
+            } else {
+                return 0;
+            }
+            break;
+        }
+    }
+    return 1;
+}
+
+static void dump_goods(figure *f)
+{
+    if (f->loads_sold_or_carrying > 0) {
+        // TODO maybe add a dump goods animation?
+        f->loads_sold_or_carrying = 0;
+        f->resource_id = RESOURCE_NONE;
+        set_cart_graphic(f);
+    }
+}
+
+static int storage_remove_resource(building *b, int resource, int amount)
+{
+    if (b->type == BUILDING_GRANARY) {
+        // For granaries, only take full units.
+        int source_amount = b->resources[resource];
+        if (source_amount < RESOURCE_ONE_LOAD) {
+            return amount;
+        }
+        return building_granary_remove_resource(b, resource, amount * RESOURCE_ONE_LOAD) / RESOURCE_ONE_LOAD;
+    } else if (b->type == BUILDING_WAREHOUSE) {
+        return building_warehouse_remove_resource(b, resource, amount);
+    } else {
+        return amount;
+    }
+}
+
+static int storage_add_resource(building *b, int resource, int amount)
+{
+    if (b->type == BUILDING_GRANARY) {
+        while (amount > 0) {
+            if (!building_granary_add_resource(b, resource, 0)) {
+                return amount;
+            }
+            amount--;
+        }
+    } else if (b->type == BUILDING_WAREHOUSE) {
+        while (amount > 0) {
+            if (!building_warehouse_add_resource(b, resource)) {
+                return amount;
+            }
+            amount--;
+        }
+    }
+    return amount;
+}
+
+void figure_depot_cartpusher_action(figure *f)
+{
+    figure_image_increase_offset(f, 12);
+    f->cart_image_id = 0;
+    int speed_factor = DEPOT_CART_PUSHER_SPEED;
+    int percentage_speed = 0;
+    int road_network_id = map_road_network_get(f->grid_offset);
+    f->terrain_usage = TERRAIN_USAGE_ROADS_HIGHWAY;
+    if (config_get(CONFIG_GP_CH_GETTING_GRANARIES_GO_OFFROAD)) {
+        f->terrain_usage = TERRAIN_USAGE_PREFER_ROADS_HIGHWAY;
+    }
+    building *b = building_get(f->building_id);
+
+    if (b->state != BUILDING_STATE_IN_USE) {
+        f->state = FIGURE_STATE_DEAD;
+        update_image(f);
+        return;
+    }
+
+    int is_linked = 0;
+    for (int i = 0; i < 3; i++) {
+        if (b->data.dock.docker_ids[i] == f->id) {
```

Comment:
Chore for the future: change `dock.docker_ids` to something like `distribution.cartpusher_ids`. Would apply to docks as well.